### PR TITLE
Add column reader writer for segment V2

### PR DESCRIPTION
**Pull request by imay**

### CODE REVIEW

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    RleDecoder<bool> null_decoder;
+    PageDecoder* data_decoder = nullptr;
+
+    // first rowid for this page
+    rowid_t first_rowid = 0;
+
+    // number of rows including nulls and not-nulls
+    uint32_t num_rows = 0;
+
+    // current offset when read this page
```
Comment (kangpinghuang) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]: Is this argument start from 0 or 1?
It seems that it start from 1 here? Add some comment to describe it?

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+                nrows_to_read -= this_run;
+                _page->offset_in_page += this_run;
+                column_view.advance(this_run);
+                _current_rowid += this_run;
+            }
+        } else {
+            RETURN_IF_ERROR(_page->data_decoder->next_batch(&nrows_to_read, &column_view));
+            DCHECK_EQ(nrows_to_read, nrows_in_page);
+
+            if (column_view.is_nullable()) {
```
Comment (kangpinghuang) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]: if the column reader is not nullable, then will column_view.is_nullable can be true?

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    Slice slices[2] = { data, Slice(checksum_buf, 4) };
+
+    bool verify_checksum = has_checksum() && _opts.verify_checksum;
+    RETURN_IF_ERROR(_file->readv_at(pp.offset, slices, verify_checksum ? 2 : 1));
+
+    if (verify_checksum) {
+        // TODO(zc): verify checksum
+    }
+
+    // TODO(zc): compress
```
Comment (kangpinghuang) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]:
```suggestion
    // TODO(zc): decompress
```

File: be/src/olap/rowset/segment_v2/encoding_info.cpp [hunk 4]
```diff
@@ -33,14 +34,33 @@ struct TypeEncodingTraits { };
 
 template<FieldType type>
 struct TypeEncodingTraits<type, PLAIN_ENCODING> {
-    static Status create_page_builder(PageBuilder** builder) {
+    static Status create_page_builder(const PageBuilderOptions& opts, PageBuilder** builder) {
         return Status::OK();
     }
-    static Status create_page_decoder(PageDecoder** decoder) {
+    static Status create_page_decoder(const Slice& data, PageDecoder** decoder) {
```
Comment (kangpinghuang) [inline be/src/olap/rowset/segment_v2/encoding_info.cpp]: Here should add PageDecoderOptions?

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "olap/rowset/segment_v2/column_reader.h"
+
+#include "env/env.h" // for RleDecoder
```
Comment (kangpinghuang) [inline be/src/olap/rowset/segment_v2/column_reader.cpp]: for RandomAccessFile?

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/rowset/segment_v2/encoding_info.h [hunk 6]
```diff
@@ -40,23 +42,28 @@ class EncodingInfo {
     // Get default type info
     static EncodingTypePB get_default_encoding_type(const TypeInfo* type_info);
 
-    Status create_page_builder(PageBuilder** builder) const {
-        return _create_buidler_func(builder);
+    Status create_page_builder(const PageBuilderOptions& opts, PageBuilder** builder) const {
+        return _create_buidler_func(opts, builder);
```
[the remark this answers is not in the document]
Comment (imay) [reply be/src/olap/rowset/segment_v2/encoding_info.h]: OK

File: be/src/olap/rowset/segment_v2/encoding_info.cpp (hunk 4 above)
Comment (imay) [reply be/src/olap/rowset/segment_v2/encoding_info.cpp]: I will add it when it is needed

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+Status ColumnReader::new_iterator(ColumnIterator** iterator) {
+    *iterator = new FileColumnIterator(this);
+    return Status::OK();
+}
+
+Status ColumnReader::read_page(const PagePointer& pp, PageHandle* handle) {
+    // Now we read this from file. we 
+    size_t data_size = pp.size;
+    if (has_checksum() && data_size < sizeof(uint32_t)) {
+        return Status::Corruption("Bad page, page size is too small");
```
[the remark this answers is not in the document]
Comment (imay) [reply be/src/olap/rowset/segment_v2/column_reader.cpp]: I think we can print log in caller

**Review by chaoyli — APPROVED** [review]

**Pull request merged**